File: frontend/src/services/product-service.js
import apiClient from './api-client';

/**
 * Product Service
 * Các API liên quan đến sản phẩm
 */
const productService = {
  /**
   * Lấy danh sách sản phẩm với filters
   * @param {Object} params - Query parameters
   * @returns {Promise}
   */
  getProducts: (params = {}) => {
    // Parse sort param into sortBy and direction for backend
    const { sort, ...otherParams } = params;
    let queryParams = { ...otherParams };
    
    if (sort) {
      const [sortBy, direction] = sort.split(',');
      queryParams.sortBy = sortBy || 'createdAt';
      queryParams.direction = direction || 'desc';
    }
    
    return apiClient.get('/products', { params: queryParams });
  },

  /**
   * Lấy sản phẩm nổi bật (Featured Products)
   * @param {Number} page - Page number (default: 0)
   * @param {Number} size - Page size (default: 8)
   * @returns {Promise}
   */
  getFeaturedProducts: (page = 0, size = 8) => {
    return apiClient.get('/products', {
      params: {
        page,
        size,
        featured: true,
      },
    });
  },

  /**
   * Lấy sản phẩm mới (New Arrivals)
   * @param {Number} page - Page number (default: 0)
   * @param {Number} size - Page size (default: 8)
   * @returns {Promise}
   */
  getNewArrivals: (page = 0, size = 8) => {
    return apiClient.get('/products', {
      params: {
        page,
        size,
        sortBy: 'createdAt',
        direction: 'desc',
      },
    });
  },

  /**
   * Lấy chi tiết sản phẩm
   * @param {Number} id - Product ID
   * @returns {Promise}
   */
  getProductById: (id) => {
    return apiClient.get(`/products/${id}`);
  },

  /**
   * Tìm kiếm sản phẩm
   * @param {String} keyword - Từ khóa tìm kiếm
   * @param {Object} params - Query parameters khác
   * @returns {Promise}
   */
  searchProducts: (keyword, params = {}) => {
    return apiClient.get('/products/search', {
      params: {
        keyword,
        ...params,
      },
    });
  },

  /**
   * Lấy sản phẩm theo category
   * @param {Number} categoryId - Category ID
   * @param {Object} params - Query parameters
   * @returns {Promise}
   */
  getProductsByCategory: (categoryId, params = {}) => {
    // Parse sort param into sortBy and direction for backend
    const { sort, ...otherParams } = params;
    let queryParams = { ...otherParams };
    
    if (sort) {
      const [sortBy, direction] = sort.split(',');
      queryParams.sortBy = sortBy || 'createdAt';
      queryParams.direction = direction || 'desc';
    }
    
    return apiClient.get(`/products/category/${categoryId}`, {
      params: queryParams,
    });
  },

  /**
   * Lấy sản phẩm đang Sale
   * @param {Object} params - Query parameters
   * @returns {Promise}
   */
  getProductsBySale: (params = {}) => {
    // Parse sort param into sortBy and direction for backend
    const { sort, ...otherParams } = params;
    let queryParams = { ...otherParams };
    
    if (sort) {
      const [sortBy, direction] = sort.split(',');
      queryParams.sortBy = sortBy || 'createdAt';
      queryParams.direction = direction || 'desc';
    }
    
    return apiClient.get('/products/sale', { params: queryParams });
  },

  /**
   * Lấy sản phẩm liên quan
   * @param {Number} productId - Product ID
   * @param {Number} limit - Số lượng sản phẩm (default: 4)
   * @returns {Promise}
   */
  getRelatedProducts: (productId, limit = 4) => {
    // Lấy sản phẩm cùng category (logic có thể tùy chỉnh)
    return apiClient.get('/products', {
      params: {
        size: limit,
      },
    });
  },
  // ============ ADMIN METHODS ============

  /**
   * Lấy tất cả sản phẩm cho admin (bao gồm inactive)
   * @param {Object} params - Query parameters
   * @returns {Promise}
   */
  getAllProductsAdmin: (params = {}) => {
    return apiClient.get('/admin/products', { params });
  },

  /**
   * Tạo sản phẩm mới
   * @param {Object} data - Product data
   * @returns {Promise}
   */
  createProduct: (data) => {
    return apiClient.post('/admin/products', data);
  },

  /**
   * Cập nhật sản phẩm
   * @param {Number} id - Product ID
   * @param {Object} data - Product data
   * @returns {Promise}
   */
  updateProduct: (id, data) => {
    return apiClient.put(`/admin/products/${id}`, data);
  },

  /**
   * Xóa sản phẩm
   * @param {Number} id - Product ID
   * @returns {Promise}
   */
  deleteProduct: (id) => {
    return apiClient.delete(`/admin/products/${id}`);
  },
};

export default productService;

File: backend/src/main/java/com/d4k/ecommerce/modules/product/service/ProductService.java
package com.d4k.ecommerce.modules.product.service;

import com.d4k.ecommerce.modules.product.dto.request.ProductRequest;
import com.d4k.ecommerce.modules.product.dto.response.ProductResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Product Service Interface
 * Định nghĩa business logic cho quản lý products
 */
public interface ProductService {
    
    /**
     * Tạo product mới (Admin)
     * @param request thông tin product
     * @return product đã tạo
     */
    ProductResponse createProduct(ProductRequest request);
    
    /**
     * Cập nhật product (Admin)
     * @param id product ID
     * @param request thông tin cần cập nhật
     * @return product sau khi cập nhật
     */
    ProductResponse updateProduct(Long id, ProductRequest request);
    
    /**
     * Xóa product (Admin)
     * @param id product ID
     */
    void deleteProduct(Long id);
    
    /**
     * Lấy chi tiết product theo ID
     * @param id product ID
     * @return thông tin product
     */
    ProductResponse getProductById(Long id);
    
    /**
     * Lấy danh sách products có phân trang (Public - chỉ active)
     * @param pageable thông tin phân trang
     * @return danh sách products
     */
    Page<ProductResponse> getAllProducts(Pageable pageable);
    
    /**
     * Lấy danh sách products theo category (Public - chỉ active)
     * @param categoryId category ID
     * @param pageable thông tin phân trang
     * @return danh sách products
     */
    Page<ProductResponse> getProductsByCategory(Long categoryId, Pageable pageable);
    
    /**
     * Lấy danh sách products đang giảm giá (Public - chỉ active)
     * @param pageable thông tin phân trang
     * @return danh sách products
     */
    Page<ProductResponse> getProductsBySale(Pageable pageable);
    
    /**
     * Tìm kiếm products theo keyword (Public - chỉ active)
     * @param keyword từ khóa tìm kiếm
     * @param pageable thông tin phân trang
     * @return danh sách products
     */
    Page<ProductResponse> searchProducts(String keyword, Pageable pageable);
    
    /**
     * Lấy tất cả products (Admin - bao gồm inactive)
     * @param pageable thông tin phân trang
     * @return danh sách products
     */
    Page<ProductResponse> getAllProductsAdmin(Pageable pageable);
}

